File: 34. Find First and Last Position of Element in Sorted Array/solution.py
class Solution(object):
    def searchRange(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: List[int]
        """
        if not nums:
            return [-1, -1]
        sol = self.search(nums, target, 0, len(nums)-1)
        if not sol:
            return [-1, -1]
        else:
            if len(sol) == 1:
                return [sol[0], sol[0]]
            else:
                return sol
        
    def search(self, nums, target, low, high):
        if high - low == 0:
            if nums[low] == target:
                return [low]
            else:
                []
        if high - low <= 1:
            sol = []
            if nums[low] == target:
                sol.append(low)
            if nums[high] == target:
                sol.append(high)
            return sol
        
        sol = []
        mid = (low+high)//2
        if nums[mid] == target:
            sol.append(mid)
            sol.extend(self.search(nums, target, low, mid-1))
            sol.extend(self.search(nums, target, mid+1, high))
            return [min(sol), max(sol)]
        elif nums[mid] < target:
            return self.search(nums, target, mid+1, high)
        else:
            return self.search(nums, target, low, mid-1)

            

File: 34. Find First and Last Position of Element in Sorted Array/solution.java
class Solution {
    public int[] searchRange(int[] nums, int target) {
        int[] sol = new int[] {-1, -1};
        if (nums.length != 0) {
            binarySearch(nums, target, sol, 0, nums.length-1);
        }
        return sol;
    }
    
    public void binarySearch(int[] nums, int target, int[] sol, int low, int high) {
        if (low == high) {
            if (nums[low] == target) {
                saveSolution(sol, low);
            }
            return;
        } else if (low > high) {
            return;
        }
        
        int mid = (low + high)/2;
        
        if (nums[mid] == target) {
            saveSolution(sol, mid);
            binarySearch(nums, target, sol, low, mid-1);
            binarySearch(nums, target, sol, mid+1, high);
        }
        else if (nums[mid] > target) {
            binarySearch(nums, target, sol, low, mid-1);
        } else {
            binarySearch(nums, target, sol, mid+1, high);
        }
    }
    
    public void saveSolution(int[] sol, int pos) {
        if (sol[0] == -1) {
            sol[0] = pos;
            sol[1] = pos;
        } else {
            if (pos < sol[0]) {
                sol[0] = pos;
            } else if (pos > sol[1]) {
                sol[1] = pos;
            }
        }
    }
}
